File: 0154-find-minimum-in-rotated-sorted-array-ii/0154-find-minimum-in-rotated-sorted-array-ii.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # l, r = 0, len(nums) - 1
        l = 0
        r = len(nums) - 1
        # while l < r and nums[l] == nums[r]: l += 1
        while l < r and nums[l] == nums[r]:
            l += 1
        while l < r:
            m = (r - l) // 2 + l
            if nums[m] > nums[r]:
                l = m + 1
            else:
                r = m
        return nums[r]
    

File: 0154-find-minimum-in-rotated-sorted-array-ii/0154-find-minimum-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int l = 0, r = nums.size() - 1;
        while(l < r && nums[l] == nums[r]) ++l;
        while(l < r) {
            int m = (r - l) / 2 + l;
            if(nums[m] > nums[r]) l = m + 1;
            else r = m;
        }
        return nums[r];
    }
};